lib/iris/util.py

```diff
@@ -2025,3 +2052,297 @@ def _strip_metadata_from_dims(cube, dims):
             reduced_cube.remove_cell_measure(cm)
 
     return reduced_cube
+
+
+class _LiftEmptyMasks:
+    """All docstring should be in _LiftEmptyMasks.decorator - public facing."""
+
+    class MaskTypes(Enum):
+        # Non-array object.
+        NOT_APPLICABLE = 0
+        # Some points actually masked.
+        SOME_TRUE = 1
+        # Mask is an array, all values False.
+        ARRAY_FALSE = 2
+        # Mask is a single value of False (np.ma.nomask).
+        SCALAR_FALSE = 3
+
+    false_mask_types = (MaskTypes.ARRAY_FALSE, MaskTypes.SCALAR_FALSE)
+
+    class InputsAddress(typing.NamedTuple):
+        """
+        Agnostic way of recording locations in both self.func_args and
+        self.func_kwargs
+        """
+
+        in_kwargs: bool
+        index_or_key: typing.Hashable or int
+
+    @staticmethod
+    def is_cube(array_or_cube):
+        from iris.cube import Cube
+
+        return isinstance(array_or_cube, Cube)
+
+    @staticmethod
+    def get_array_lib(data):
+        if iris._lazy_data.is_lazy_data(data):
+            array_lib = da
+        else:
+            array_lib = np
+        return array_lib
+
+    @classmethod
+    def get_data(cls, array_or_cube):
+        if cls.is_cube(array_or_cube):
+            result = array_or_cube.core_data()
+        elif hasattr(array_or_cube, "__array__"):
+            # Generic check for NumPy, Dask and masked arrays.
+            result = array_or_cube
+        else:
+            result = None
+
+        return result
+
+    @classmethod
+    def get_mask_info(cls, array_or_cube) -> Tuple[MaskTypes, Number]:
+        data = cls.get_data(array_or_cube)
+        if data is not None and iris._lazy_data.is_lazy_data(data):
+            # Take a sample to check the mask - dask.array.ma arrays have
+            #  no distinguishing properties.
+            sample = iris._lazy_data.sample(data)
+        else:
+            sample = data
+
+        if data is not None and hasattr(sample, "mask"):
+            if is_masked(data):
+                mask_type = cls.MaskTypes.SOME_TRUE
+            else:
+                if sample.mask is ma.nomask:
+                    mask_type = cls.MaskTypes.SCALAR_FALSE
+                else:
+                    mask_type = cls.MaskTypes.ARRAY_FALSE
+        else:
+            mask_type = cls.MaskTypes.NOT_APPLICABLE
+
+        fill_value = getattr(sample, "fill_value", None)
+
+        return mask_type, fill_value
+
+    def __init__(self, func_name: str, *func_args, **func_kwargs):
+        from iris.cube import CubeList
```

Comment:
I'm used to seeing the `__init__` at the top of a class, I don't believe there is a PEP prescribing method order, but this stackoverflow answer gives a good rationale for this ordering https://stackoverflow.com/a/36967648.

Reply:
Sounds good, looks like I got my `classmethod` and `staticmethod` ordering the wrong way round

Reply:
c3528b560